File: app/src/main/java/org/eyeseetea/malariacare/domain/usecase/GetServerUseCase.java
package org.eyeseetea.malariacare.domain.usecase;

import org.eyeseetea.malariacare.domain.boundary.executors.IAsyncExecutor;
import org.eyeseetea.malariacare.domain.boundary.executors.IMainExecutor;
import org.eyeseetea.malariacare.domain.boundary.repositories.IServerRepository;
import org.eyeseetea.malariacare.domain.entity.Server;

public class GetServerUseCase implements UseCase{

    public interface Callback{
        void onSuccess(Server server);
    }

    private IServerRepository serverRepository;
    private IMainExecutor mainExecutor;
    private IAsyncExecutor asyncExecutor;
    private Callback mCallback;

    public GetServerUseCase(
            IServerRepository serverRepository,
            IMainExecutor mainExecutor,
            IAsyncExecutor asyncExecutor) {
        this.mainExecutor = mainExecutor;
        this.asyncExecutor = asyncExecutor;
        this.serverRepository = serverRepository;
    }

    public void execute(Callback callback) {
        mCallback = callback;
        asyncExecutor.run(this);
    }

    @Override
    public void run() {
        try {
            notifyComplete(serverRepository.getGetLoggedServer());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void notifyComplete(Server server) {
        mainExecutor.run(() -> mCallback.onSuccess(server));
    }

}

File: app/src/main/java/org/eyeseetea/malariacare/domain/usecase/GetServerUseCase.kt
package org.eyeseetea.malariacare.domain.usecase

import org.eyeseetea.malariacare.domain.boundary.executors.IAsyncExecutor
import org.eyeseetea.malariacare.domain.boundary.executors.IMainExecutor
import org.eyeseetea.malariacare.domain.boundary.repositories.IServerRepository
import org.eyeseetea.malariacare.domain.common.Either
import org.eyeseetea.malariacare.domain.entity.Server

sealed class GetServerFailure {
    object ServerNotFoundFailure : GetServerFailure()
}

// TODO: this use case should not use executors, we must use sync version
// because presenters should execute use case using executors
class GetServerUseCase(
    private val serverRepository: IServerRepository,
    private val mainExecutor: IMainExecutor,
    private val asyncExecutor: IAsyncExecutor
) : UseCase {
    interface Callback {
        fun onSuccess(serverResult: Either<GetServerFailure, Server>)
    }

    private lateinit var mCallback: Callback

    fun execute(callback: Callback) {
        mCallback = callback
        asyncExecutor.run(this)
    }

    override fun run() {
        notifyComplete(serverRepository.getLoggedServer())
    }

    private fun notifyComplete(serverResult: Either<GetServerFailure, Server>) {
        mainExecutor.run { mCallback.onSuccess(serverResult) }
    }
}
